### ISSUE TITLE
Ego car predefine routing actions cannot be corrected by Esmini ALKS controller:

### ISSUE THREAD

**Issue by Iqbalparvi** [esmini/esmini#304]
We are trying to do some experiments on Esmini (which could be an extension of our previous research paper: https://ieeexplore.ieee.org/document/9477676). 
At the esmini platform, the yellow line in the below figure shows the trajectories/waypoints with unintentional lane departures. 
![image](https://user-images.githubusercontent.com/47267894/178652750-b05f5f43-c31b-449b-a5dc-774fb3a6dc47.png)
Our console output shows that we successfully activated the ALKS controller. 
![image](https://user-images.githubusercontent.com/47267894/178652871-586eebe5-5515-4b42-8767-5d532415ba0d.png)

However, in the OSC window, the controller performance is not as expected as it is not controlling the ego car from the lane departures.
 It seems that the esmini ignores the predefine routing action inside the Open Scenarios during simulation, which is an unintentional lane departure in our scenarios. 
**Could you please have a look at our experiment proposal, and see if it is achievable by using Esmini and OpenX? Or could it be that this is a bug of esmini?**
**Proposal:** We want to repeat the demo of MATLAB LKAS, where a predefined careless ego-car trajectory can be corrected in the simulation. 
To achieve this on Esmini (in terms of OpenX format), we have already had some careless ego-car trajectories to be followed by the ego car (by FollowTrajectoryAction), we consider our experiment lacks an LKAS-like controller that can correct the ego-car's careless trajectory (by applying lateral control to the ego car in parallel with FollowTrajectoryAction). 
The below figure demonstrates how the MATLAB LKAS works. 
![Simulink_LKAS](https://user-images.githubusercontent.com/47267894/178653000-581563b6-ca48-466c-beb7-dbb7e4d2a20e.jpeg)
The blue line shows the ego-car's careless trajectory in the design phase, and the red line shows the LKAS corrects the ego-car' driving trajectory in actual simulation, which keeps the ego-car to drive in its intended lane. 
Please have a look at our demo OpenX files, it shows our idea, but the ego-car's careless trajectory cannot be corrected yet.
[LKA_Dashed_line_solid_left.zip](https://github.com/esmini/esmini/files/9098993/LKA_Dashed_line_solid_left.zip)
Many thanks for your help!

**Comment by eknabevcc** [comment]
esmini, and perhaps OpenSCENARIO, is perhaps a bit un-flexible for this use case. What happens in your example is that esmini tries to operate the ALKS controller in addition to any scenario events. That's why the followTrajectory action has effect while/despite the controller being active. 

The reason is the esmini specific controller `mode` attribute which can be `additive` or `override`. If set to `additive` (which happens to be the default for ALKS controller) the default controller will be incorporated (executing any events specified in the scenario). If mode = `override`, the default controller will be skipped, i.e. any scenario events will be ignored for that object (and for the relevant domain, of course).

So, potentially the solution for your use case would be to set the mode = `override´, as in attached modified scenario: [LKA_Dashed_line_solid_left-mod.zip](https://github.com/esmini/esmini/files/9261150/LKA_Dashed_line_solid_left-mod.zip). Unfortunately I don't think it helps your case since you probably expect the controller to mitigate or take counter actions to the lateral motion of the followTrajectory action when it gets too "bad". However, the ALKS controller is much simpler: It will just accept and keep current lane offset. 

To summarize: With mode=additive the lateral position from trajectory will be accepted by the ALKS controller and with mode=override it (the trajectory) will be totally ignored. I don't see any easy or quick fix for your use case unfortunately. It would require an improvement of the ALKS controller so that lateral lane offset would be not just accepted but input to a dynamic steering control.

**Comment by Iqbalparvi** [comment]
Hi @eknabevcc;
Thank you so much for your response to our discussion.
As mentioned above, "**It seems that the Esmini ignores the predefine routing action inside the Open Scenarios during simulation”.**
Our exploration seems very beneficial to enhancing the Esmini simulator’s robustness in ADAS with an Open Scenario. A mechanism to control the ego-car careless trajectory and to differentiate the user define and controller performance actions will improve the performance of the ALKS controller to simulate further scenarios.
Could you please confirm that:

- Either is it a bug in Esmini where the situation is possible but never tested previously?
- Is it possible the LKA Controller will be implemented in the future?
Regards

**Comment by eknabevcc** [comment]
Just to clarify: Your statement "...esmini ignores the predefined routing action..." means: esmini DO perform the routing action but the ALKS controller do no counter action to prevent that action. This is how I now understand it anyway :)

In that case you're right. The ALKS controller is not yet capable of doing active steering to keep in lane center. It will only keep current lane offset, whatever it is. In this case it will be a result of applying the trajectory action.

It might be that the esmini ALKS controller will be updated in future to also do dynamic steering (LKA). But still it's not super clear to me whether the scenario events should have effect or not. I mean, a controller takes over control for a specified domain(s) (lateral and/or longitudinal). It's up to the controller whether to consider scenario events for the entity or not. But here the esmini controller attribute `mode` (additive/override) could perhaps be used to control exactly that behavior, should the scenario events be considered or not.

I totally agree that your use case is interesting and useful. So, just for reasoning: What would the expected behavior be in your (scenario) case? Is it the following: As soon as the lane offset pass a threshold value the FollowTrajectoryAction is cancelled completely. Or would it be an ongoing "fight" between the trajectory action and the ALKS controller. Which would be a bit strange, since the vehicle will move away from the trajectory (at least if ALKS is the stronger power) - which potentially would result in a "driver" trying to steer increasingly away from lane, trying to reach the trajectory...

What is your expectations and thinking? Have you run the scenario in another simulator tool, what was the behavior then?

**Comment by Iqbalparvi** [comment]
Thank you so much for your response to this use case.
1.	We totally agreed with your statement **“Esmini DO perform the routing action but the ALKS controller do no counter action to prevent that action”**,  and that is the reason we have mentioned the **ALKS controller** in the subject area.
2.	**What would the expected behavior be in your (scenario) case? Is it the following: As soon as the lane offset pass a threshold value the FollowTrajectoryAction is cancelled completely.**.
Yes, exactly our expected behavior is that when the careless trajectory occurs the LKAS controller should correct the lane departures to keep the Ego vehicle in lane.
It assumes that when the lane departure is about to occur and no further action is taken from the drive side, LKAS will apply to assist torque to the steer to prevent the lane departures.
3.	**Have you run the scenario in another simulator tool, what was the behavior then?**

 Yes, we have run the scenario with different Open and closed-sourced simulator tools/platforms.
One of them is MATLAB/Simulink, where the prebuild ADAS’s Simulink Testbench modules are available to simulate and verify different scenarios.
The LKAS functions in this platform can understand more clearly from the below figure.
![comparison](https://user-images.githubusercontent.com/47267894/191433922-ef92fa64-ed07-4476-9146-1c493128ea0f.png)

In the left figure, the blue line shows the ego-car careless trajectory in the design phase, and the red line shows the Simulink  LKAS controller corrects the ego-car' driving trajectory in actual simulation, which keeps the ego-car driving in its intended lane.
To achieve this on Esmini when execute the scenario (in terms of OpenX format), the ALKS controller performance is not the same as expected.
The next(right) figure shows the ALKS controller performance on Esmini.
The yellow line represents the ego car's careless trajectories as defined in the scenario, which is expected to be corrected by the ALKS controller by following the (assumed) red line to keep the vehicle in the lane. 
Also, the expected behavior is observed when the scenarios are run on other simulators/platforms to test their embedded LKASs performance.

**Comment by eknabevcc** [comment]
Thanks for clarifying the expectations. It's a good use case, I must agree. 

To conclude, this feature probably requires some work in esmini why I can't provide a fix right now. But I will keep it in mind and hopefully this feature will be prioritized in the backlog. 

Thanks again for the input!

**Comment by Iqbalparvi** [comment]
We really thank you for this conclusion and appreciate your time and effort.
Best wishes.
